panic: runtime error: slice bounds out of range [:-6] after upgrade v1.0

**Describe the bug**

I upgraded tempo to v1.0 (thanks for the release, amazing product 👍). First boot after upgrade was fine. Now I'm having this error en restart and I can no longer start tempo:

```
[...]
level=info ts=2021-06-10T08:05:43.391436169Z caller=main.go:94 msg="Starting Tempo" version="(version=1.0.0, branch=HEAD, revision=b92405378)"
level=info ts=2021-06-10T08:05:43.392640963Z caller=server.go:239 http=[::]:3100 grpc=[::]:9095 msg="server listening on addresses"
level=info ts=2021-06-10T08:05:44.310411806Z caller=frontend.go:29 msg="creating tripperware in query frontend to shard queries"
ts=2021-06-10T08:05:44Z level=info msg="OTel Shim Logger Initialized" component=tempo
level=warn ts=2021-06-10T08:05:44.312534002Z caller=modules.go:145 msg="Worker address is empty in single binary mode. Attempting automatic worker configuration. If queries are unresponsive consider configuring the worker explicitly." address=127.0.0.1:9095
level=info ts=2021-06-10T08:05:44.312759423Z caller=worker.go:113 msg="Starting querier worker connected to query-frontend" frontend=127.0.0.1:9095
ts=2021-06-10T08:05:44Z level=info msg="No sampling strategies provided or URL is unavailable, using defaults" component=tempo
ts=2021-06-10T08:05:44Z level=info msg="Starting GRPC server on endpoint 0.0.0.0:55680" component=tempo
level=info ts=2021-06-10T08:05:44.313354175Z caller=module_service.go:59 msg=initialising module=store
level=info ts=2021-06-10T08:05:44.313415725Z caller=module_service.go:59 msg=initialising module=server
level=info ts=2021-06-10T08:05:44.313576065Z caller=module_service.go:59 msg=initialising module=memberlist-kv
level=info ts=2021-06-10T08:05:44.313622115Z caller=module_service.go:59 msg=initialising module=query-frontend
level=info ts=2021-06-10T08:05:44.313620395Z caller=module_service.go:59 msg=initialising module=overrides
level=info ts=2021-06-10T08:05:44.313696956Z caller=module_service.go:59 msg=initialising module=compactor
level=info ts=2021-06-10T08:05:44.313703076Z caller=module_service.go:59 msg=initialising module=ring
level=info ts=2021-06-10T08:05:44.313719756Z caller=compactor.go:98 msg="waiting for compaction ring to settle" waitDuration=1m0s
level=info ts=2021-06-10T08:05:44.313804716Z caller=ring.go:271 msg="ring doesn't exist in KV store yet"
level=info ts=2021-06-10T08:05:44.313837016Z caller=module_service.go:59 msg=initialising module=ingester
level=info ts=2021-06-10T08:05:44.313840656Z caller=client.go:247 msg="value is nil" key=collectors/ring index=1
level=info ts=2021-06-10T08:05:44.313852036Z caller=ingester.go:320 msg="beginning wal replay"
level=info ts=2021-06-10T08:05:44.313899407Z caller=module_service.go:59 msg=initialising module=querier
level=info ts=2021-06-10T08:05:44.313950537Z caller=module_service.go:59 msg=initialising module=distributor
level=info ts=2021-06-10T08:05:44.313959637Z caller=wal.go:93 msg="beginning replay" file=521ddd7b-d4d9-4e60-8272-85842bacefcc:single-tenant:v2:none:v1 size=3690496
level=info ts=2021-06-10T08:05:44.314419638Z caller=worker.go:193 msg="adding connection" addr=127.0.0.1:9095
panic: runtime error: slice bounds out of range [:-6]

goroutine 423 [running]:
github.com/grafana/tempo/tempodb/encoding/v2.unmarshalPageFromReader(0x29277a0, 0xc000121548, 0x29556f0, 0x39af8a0, 0xc040dfe280, 0x25a, 0x25f, 0x10, 0xc0417ff150, 0xc00007a070)
/drone/src/tempodb/encoding/v2/page.go:84 +0x428
github.com/grafana/tempo/tempodb/encoding/v2.(*dataReader).NextPage(0xc040e70d40, 0x0, 0x0, 0x0, 0x0, 0xc0417ff150, 0x10, 0x10, 0x0, 0x10)
/drone/src/tempodb/encoding/v2/data_reader.go:71 +0x9d
github.com/grafana/tempo/tempodb/encoding/v1.(*dataReader).NextPage(0xc040e7fb00, 0x0, 0x0, 0x0, 0x10, 0xc0417ff150, 0x0, 0x10, 0x0, 0x2920ea0)
/drone/src/tempodb/encoding/v1/data_reader.go:85 +0x5e
github.com/grafana/tempo/tempodb/wal.newAppendBlockFromFile(0xc040e123cf, 0x3d, 0xc000143010, 0xe, 0x0, 0x0, 0x0, 0x0, 0x0)
/drone/src/tempodb/wal/append_block.go:107 +0x3bc
github.com/grafana/tempo/tempodb/wal.(*WAL).RescanBlocks(0xc00073e9a0, 0x2922800, 0xc000363ef0, 0x2, 0x0, 0x0, 0xc040f0a720, 0xc040e7f500)
/drone/src/tempodb/wal/wal.go:94 +0x425
github.com/grafana/tempo/modules/ingester.(*Ingester).replayWal(0xc0008a5600, 0x7fd89d59a2e8, 0x10)
/drone/src/modules/ingester/ingester.go:322 +0x177
github.com/grafana/tempo/modules/ingester.(*Ingester).starting(0xc0008a5600, 0x29669c0, 0xc040e70d00, 0xc040f0c001, 0xc040e6f9c0)
/drone/src/modules/ingester/ingester.go:98 +0x2f
github.com/cortexproject/cortex/pkg/util/services.(*BasicService).main(0xc000940f00)
/drone/src/vendor/github.com/cortexproject/cortex/pkg/util/services/basic_service.go:157 +0x46a
created by github.com/cortexproject/cortex/pkg/util/services.(*BasicService).StartAsync.func1
/drone/src/vendor/github.com/cortexproject/cortex/pkg/util/services/basic_service.go:119 +0xb3
```

**To Reproduce**
Steps to reproduce the behavior:
1. Start Tempo

**Expected behavior**
<!-- A clear and concise description of what you expected to happen. -->

No crash

**Environment:**
 - Infrastructure: Kubernetes with latest helm chart
 - Deployment tool: helm

**Additional Context**

I increased `grpc_server_max_recv_msg_size: 41943040` on server config, not sure it's related.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,6 +1,7 @@
 ## main / unreleased
 
 * [FEATURE] Added the ability to hedge requests with all backends [#750](https://github.com/grafana/tempo/pull/750) (@joe-elliott)
+* [BUGFIX] Guard against negative dataLength [#763](https://github.com/grafana/tempo/pull/763) (@joe-elliott)
 
 ## v1.0.0
 

--- a/integration/microservices/docker-compose.yaml
+++ b/integration/microservices/docker-compose.yaml
@@ -16,6 +16,7 @@ services:
   ingester-0:
     image: tempo:latest
     command: "-target=ingester -config.file=/etc/tempo.yaml"
+    restart: always
     volumes:
       - ./tempo.yaml:/etc/tempo.yaml
     ports:
@@ -28,6 +29,7 @@ services:
   ingester-1:
     image: tempo:latest
     command: "-target=ingester -config.file=/etc/tempo.yaml"
+    restart: always
     volumes:
       - ./tempo.yaml:/etc/tempo.yaml
     ports:
@@ -40,6 +42,7 @@ services:
   ingester-2:
     image: tempo:latest
     command: "-target=ingester -config.file=/etc/tempo.yaml"
+    restart: always
     volumes:
       - ./tempo.yaml:/etc/tempo.yaml
     ports:
@@ -50,7 +53,7 @@ services:
       - distributor # inverted relationship here to add delay for minio
 
   minio:
-    image: minio/minio:RELEASE.2021-05-26T00-22-46Z
+    image: minio/minio:RELEASE.2021-06-14T01-29-23Z
     environment:
       - MINIO_ACCESS_KEY=tempo
       - MINIO_SECRET_KEY=supersecret

--- a/tempodb/encoding/v2/page.go
+++ b/tempodb/encoding/v2/page.go
@@ -79,7 +79,7 @@ func unmarshalPageFromReader(r io.Reader, header pageHeader, buffer []byte) (*pa
 	dataLength := int(totalLength) - totalHeaderSize
 
 	if dataLength < 0 {
-		return nil, fmt.Errorf("Unexpected negative dataLength unmarshalling page: %d", dataLength)
+		return nil, fmt.Errorf("unexpected negative dataLength unmarshalling page: %d", dataLength)
 	}
 
 	if cap(buffer) < dataLength {

--- a/tempodb/encoding/v2/page.go
+++ b/tempodb/encoding/v2/page.go
@@ -78,6 +78,10 @@ func unmarshalPageFromReader(r io.Reader, header pageHeader, buffer []byte) (*pa
 	}
 	dataLength := int(totalLength) - totalHeaderSize
 
+	if dataLength < 0 {
+		return nil, fmt.Errorf("Unexpected negative dataLength unmarshalling page: %d", dataLength)
+	}
+
 	if cap(buffer) < dataLength {
 		buffer = make([]byte, dataLength)
 	} else {

--- a/tempodb/encoding/v2/page_test.go
+++ b/tempodb/encoding/v2/page_test.go
@@ -16,7 +16,7 @@ type testHeader struct {
 
 func (h *testHeader) unmarshalHeader(b []byte) error {
 	if len(b) != int(uint32Size) {
-		return errors.New("err")
+		return errors.New("testheader err")
 	}
 
 	h.field = binary.LittleEndian.Uint32(b[:uint32Size])
@@ -107,4 +107,27 @@ func TestPageMarshalToBuffer(t *testing.T) {
 		assert.Equal(t, uint32(len(outputBuffer)), page.totalLength)
 		assert.Equal(t, tc.field, page.header.(*testHeader).field)
 	}
+}
+
+func TestCorruptHeader(t *testing.T) {
+	data := []byte{0x01, 0x02, 0x03}
+	buff := &bytes.Buffer{}
+
+	bytesWritten, err := marshalPageToWriter(data, buff, constDataHeader)
+	require.NoError(t, err)
+	assert.Equal(t, len(data)+constDataHeader.headerLength()+int(baseHeaderSize), bytesWritten)
+
+	buffBytes := buff.Bytes()
+
+	// overwrite the base header with 0s
+	zeroHeader := bytes.Repeat([]byte{0x00}, baseHeaderSize)
+	copy(buffBytes, zeroHeader)
+
+	page, err := unmarshalPageFromBytes(buffBytes, constDataHeader)
+	assert.Nil(t, page)
+	assert.EqualError(t, err, "expected data len -6 does not match actual 3")
+
+	page, err = unmarshalPageFromReader(bytes.NewReader(buffBytes), constDataHeader, nil)
+	assert.Nil(t, page)
+	assert.EqualError(t, err, "unexpected negative dataLength unmarshalling page: -6")
 }

